### PR TITLE
legder-api: add flag to prune all divulged contracts

### PR DESCRIPTION
**Pull request by meiersi-da**
Context: draft PR to discuss Ledger API changes to allow plugging the storage leak from not pruning unarchived, but stale divulged contracts in a backwards-compatible way.

TODO:
- [ ] complete design doc that this PR is part of
- [ ] move changelog entry from below to commit

CHANGELOG_BEGIN
Ledger API Server: pruning takes a new flag to force it to delete all divulged contracts up to the pruning offset instead of only the ones for which there is an consuming exercise event before the pruning offset. 

We recommend always setting this flag when pruning participant nodes whose applications do not rely on divulgence, which should be the default for newly built applications. Otherwise, storage can be leaked on participant nodes that see a divulged contract, but not its consuming exercise. 

For example in the second scenario of https://docs.daml.com/concepts/ledger-model/ledger-privacy.html#divulgence-when-non-stakeholders-see-contracts, the creation of `Iou $Bank P` is visible to A's node, but not its consumption by `P`s request to settle it. Thus, the create event for `Iou $Bank P` would never get pruned on A's node unless the flag to prune all divulged events is set on the pruning request to A's node.

Note that using divulged contracts as part of command interpretation is deprecated. Once support for divulged contracts has been removed, this flag can always be set to true without impairing the Daml applications that run on a participant node.
CHANGELOG_END

FYI: @bame-da @matthiasS-da @mziolekda @tudor-da @gerolf-da 

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by meiersi-da** [comment]
Part of https://github.com/digital-asset/daml/issues/9283

**Review by gerolf-da — COMMENTED** [review]
Thank you, @meiersi-da. This is a nice way to remain backwards compatibility and opt-in to the stricter pruning functionality during the deprecation period of "usage of divulged contracts as inputs to transactions".
Would you say that `prune_all_divulged_contracts=true` will be the default pruning behavior, once the deprecated feature is removed? If yes, do you think it would be useful to state this in the docs as "preview"?

**Comment by matthiasS-da** [comment]
Thanks for starting this!

I suppose you will pass the new flag through the write service so that the ledger implementation knows what to prune?

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/participant_pruning_service.proto [hunk 1]
```diff
@@ -39,9 +39,27 @@ service ParticipantPruningService {
...
 message PruneRequest {
-  // Inclusive offset up to and at which the ledger is to be pruned.
+  // Inclusive offset up to which the ledger is to be pruned.
+  // By default the following data is pruned:
+  //   1. All normal and divulged contracts that have been archived before
+  //   `  prune_up_to`.
+  //   2. All transaction events and completions before `prune_up_to`
   string prune_up_to = 1;
 
+  // Prune all divulged contracts created before `prune_up_to` independent of
```
[the remark this answers is not in the document]
Comment (meiersi-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/participant_pruning_service.proto:49]: I'm writing the docs from the perspective of somebody knowing only about the ledger model. I'm not sure the concept of events is clear there. I agree though that formulation needs to improve. What do you think about the following:
```
   // Prune all contract divulgence information for contracts divulged before `prune_up_to` 
```

**Comment by meiersi-da** [comment]
> Thank you, @meiersi-da. This is a nice way to remain backwards compatibility and opt-in to the stricter pruning functionality during the deprecation period of "usage of divulged contracts as inputs to transactions".
> Would you say that `prune_all_divulged_contracts=true` will be the default pruning behavior, once the deprecated feature is removed? If yes, do you think it would be useful to state this in the docs as "preview"?

I would expect that this will become the default at some point in time. I would though not document that on the API docs, as they should just document the status quo. I'd rather add it to the Daml roadmap, and potentially to the release notes where it can be explained in the context of the goal of removing divulgence.

**Comment by meiersi-da** [comment]
> Thanks for starting this!
> 
> I suppose you will pass the new flag through the write service so that the ledger implementation knows what to prune?

Yes, good point. We'll do that. What do you intend Canton to do in case it is told to not prune all divulged contracts?

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/participant_pruning_service.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (gerolf-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/participant_pruning_service.proto:49]: I like the new phrasing.

> I'm writing the docs from the perspective of somebody knowing only about the ledger model. I'm not sure the concept of events is clear there.

I see. In this context though (Ledger API docs), the concept of events, event witnesses, etc. is rather well defined, and ultimately what the participant operator cares about. Regardless, the new phrasing covers all aspects by being less specific.

**Comment by matthiasS-da** [comment]
If Canton is told to not prune all divulged contracts, it will only prune archived contracts (up to the specified global offset).

**Comment by meiersi-da** [comment]
Fixed as part of #10635

**Pull request closed without merging**